### ISSUE TITLE
Bower dependencies might import a second angular version

### ISSUE THREAD

**Issue by Xample** [Urigo/angular-meteor#205]
Hi I just got the "WARNING: Tried to load angular more than once." message into the console.
This happens when I do add a dependency into bower which one relies on a latest version of angular.

```bash
 meteor create test
 cd test/
 meteor add mquandalle:bower
 meteor add urigo:angular
 meteor
```
Create a bower.json file with this content:
```json
{
  "name": "my-app",
  "version": "0.0.1",
  "dependencies": {
    "angulartics": "0.17.2"
  },
  "private": true
}
```
Then run the project
```bash
meteor
```
Into the terminal I get
```bash
=> Started proxy.                             
Bower:  angulartics v0.17.2 successfully installed
Bower:  angular v1.3.14 successfully installed
Bower:  waypoints v2.0.5 successfully installed
Bower:  SHA-1 v0.1.1 successfully installed
Bower:  jquery v2.1.3 successfully installed
=> Started MongoDB.                           
=> Started your app.                          

=> App running at: http://localhost:3000/
```

As a proof another angular has been installed, the problem is the original one from urigo:angular still seems to be there…

Into Chrome or safari, the console displays WARNING: Tried to load angular more than once. as "expected"

**Comment by Urigo** [comment]
Can you share the repo?
The problem is that probably angulartics brings angular from bower and angular-meteor now brings angular from the angularjs:angular Meteor package.

We will continue to do that until Bower will support constraint solver.

I think the best way is to create angulartics Meteor package.
Please look at here:
https://github.com/MeteorCommunity/discussions/issues/14

**Comment by Xample** [comment]
@Urigo the few steps above are here to reproduce the issue from scratch. I can make a repo if you need to. Btw I tried to update the ionic example https://github.com/netanelgilad/meteor-ionic-example just running
```bash
meteor update
```
and get the same issue.
I also updated another project of mine (not ionic), after several refactoring to meet the new specification it slowly starts to work. Still I also have 2 times this message appearing into the console.

**Comment by Urigo** [comment]
@Xample that is really strange.

Please attach a repo as I really think it's a specific problem on your repo

**Comment by renato** [comment]
I had this same error - angular-meteor gets angular from angularjs:angular Meteor package while some bower packages have angular as dependency, causing both `/packages/angularjs_angular.js` and `/packages/bower/angular/angular.js` to be added to the HTML.

My workaround is using an override in my bower.json:
```
  "overrides": {
    "angular": {
      "main": []
    }
  }
```

Meteor-bower installs angular in `.meteor/local/bower` but at least doesn't add it to the HTML.

**Comment by netanelgilad** [comment]
@Urigo @Xample about the corodova app initialization. From what I understood from angular code, they bootstrap the application using ng-app on the document ready event. That is not good for cordova apps. On cordova the correct event to listen on is deviceready. Check out this link http://stackoverflow.com/questions/21556090/cordova-angularjs-device-ready.
That is why in meteor-ionic the if(Meteor.isCordova) is there. Then, when using angular on cordova do not use the ng-app attribute.

**Comment by netanelgilad** [comment]
The two angular scripts problem is indeed causes by a bower package requesting angular as well as angular-meteor.
@renato's solution is a good workaround for now. The problem is a limitation with meteor-bower package. When angular-meteor worked with meteor-bower as well the problem still occured. That is because every package and the app itself installs its own bower_components, having angular loaded more than once.
I've been thinking about how to solve this problem with the meteor bower package but currently i'm out of luck.. maybe i'll get back to working on this issue soon.
@Urigo as I can't see a solution right now (but maybe you have something in mind?) I suggest considering documenting the workaround somewhere.

**Comment by renato** [comment]
@netanelgilad, there are two open issues in meteor-bower that concerns this issue: mquandalle/meteor-bower#21 and mquandalle/meteor-bower#64.

Do you mean that angular-meteor used bower before and this same problem occured? Is there any additional reason to not use meteor-bower?

**Comment by Urigo** [comment]
There are a few reasons why not to use meteor-bower library.
The main one is that it doesn't have a constraint solver.
There are talks about adding the new Meteor constraint solver into Bower itself, but that will take time.

The best solution for now, is to create a Meteor package that reflects the versioning of the 3rdParty one.
https://github.com/MeteorPackaging
https://github.com/MeteorPackaging/discussions

We already done that for []AngularJS](https://github.com/MeteorPackaging/angular.js), [ui-router](https://github.com/MeteorPackaging/ui-router), [angular-sanitize](https://github.com/MeteorPackaging/angular-sanitize), [angular-animate](https://github.com/MeteorPackaging/angular-animate) and angular-google-maps (https://atmospherejs.com/angularui/angular-google-maps, https://github.com/angular-ui/angular-google-maps/blob/master/package.js, https://github.com/angular-ui/angular-google-maps/pull/1174)

I know it's annoying, but meteor-bower is really not production ready and I think the 3rdParty solution is not bad

**Comment by Urigo** [comment]
I've closed it as this is for now the official solution.
When meteor-bower will get better, we might consider going back to it and recommending it for others.
The workaround is nice but there are more problems so I don't want to document this as we don't recommend it.

**Comment by Xample** [comment]
@Urigo agreed, I will probably add the dependencies manually for now. I am not again the idea of creating a package but I need to learn a little more about it (and obviously to then take time to maintain).
Thank you all for your feedback regarding this issue.

**Comment by Namek** [comment]
another workaround:
```json
{
  "ignoredDependencies": ["jquery", "angular"]
}
```

**Comment by MurWade** [comment]
@Namek  adding ignoredDependencies to the bower.json 

i get 

    While building the application:
    error: Syntax error in bower.json
    packages/bower/plugin/handler.js:18:1: Cannot read property 'dependencies' of undefined (compiling        bower.json)
    at bowerHandler (packages/bower/plugin/handler.js:18:1)
    at Package (packages/bower/plugin/handler.js:262:1)

**Comment by mgrivera** [comment]
Hi, 

this is my bower.json file: 
```javascript
{
  "name": "scrwebM",
  "version": "0.0.1",
  "dependencies": {
    "angular-ui-grid": "3.0.5"
  },
  "private": true,
  "ignoredDependencies": [
      "angular"
  ],
  "overrides": {
      "angular-ui-grid": {
          "arch": [ "client" ]
        }
    }
}
```
Even if I stated that angular should not be downloaded, it was. I saw the message from bower when running my meteor app: 

*bower: angular v1.4.8 successfully installed*

I can see it was downloaded at: *myApp/.meteor/local/bower/angular* directory.

The message: *Tried to load angular more than once* is now gone because, I think, angular (from bower) is not reached at this directory inside ".meteor". So, my app is seeing just one angular, from *angular (v1.2.2) meteor package*.

So, my question is: why was angular downloaded, as said, if I used *ignoredDependencies* in my bower.json file? 

Thanks and bye ...

**Comment by Namek** [comment]
@mgrivera why do you even care? `.meteor/local` shouldn't be in your repo anyway.

But if you're asking I've done a small research. That may be a bug of bower itself: https://github.com/bower/bower/pull/1394 But by looking at https://github.com/mquandalle/meteor-bower/blob/master/package.js it seems there's bower `1.5.2` while the feature landed in `1.6.2` (look at the last comment in issue 1394). But it all doesn't matter because Meteor packages have to support files in it's own predicted way. That said, `meteor-bower` have just implemented the feature for Meteor on it's own: https://github.com/mquandalle/meteor-bower/blob/master/plugin/handler.js#L78

**Comment by mgrivera** [comment]
@Namek Wow, thanks for your comments on this issue, they are much appreciated. I already thought that if angular was not there, ui-grid was going to complain about it and produce loading errors because, in the end, maybe it was going to load first than angular from *angular package* and was going to expect angular being there. So, I'm going to just let angular (from *bower*) be there and forget the issue. 

Thanks again and bye ...
